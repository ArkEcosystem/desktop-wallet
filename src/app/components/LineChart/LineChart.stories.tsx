import React from "react";

import { LineChart } from "./LineChart";

export default {
	title: "Components / Line Chart",
};

export const Default = () => {
	const data = [
		{ name: "Mon", label: "22 Jun, 2020", usd: 1000, btc: 2400, formatted: { usd: "1,000", btc: "0.26" } },
		{ name: "Tue", label: "23 Jun, 2020", usd: 3000, btc: 5400, formatted: { usd: "3,000", btc: "0.58" } },
		{ name: "Wed", label: "24 Jun, 2020", usd: 15100, btc: 1398, formatted: { usd: "15,100", btc: "0.15" } },
		{ name: "Thu", label: "25 Jun, 2020", usd: 4000, btc: 9800, formatted: { usd: "4,000", btc: "1.06" } },
		{ name: "Fri", label: "26 Jun, 2020", usd: 11200, btc: 2000, formatted: { usd: "11,200", btc: "0.22" } },
		{ name: "Sat", label: "27 Jun, 2020", usd: 3000, btc: 4800, formatted: { usd: "3,000", btc: "0.52" } },
		{ name: "Sun", label: "28 Jun, 2020", usd: 8000, btc: 18000, formatted: { usd: "8,000", btc: "2.16" } },
	];

	const lines = [
		{
			dataKey: "btc",
			label: "BTC",
			color: "warning-600",
		},
		{
			dataKey: "usd",
			label: "USD",
			color: "success-600",
		},
	];
	const period = "22 Jun - 28 Jun";
	return (
		<div>
			<LineChart period={period} data={data} lines={lines} onPeriodClick={() => alert("Clicked period change")} />
		</div>
	);
};
