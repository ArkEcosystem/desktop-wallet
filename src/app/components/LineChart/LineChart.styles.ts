export const chartStyles = `
	.recharts-surface,
	.recharts-cartesian-grid {
		overflow: visible;
	}

    .recharts-cartesian-axis-tick-line {
        stroke: currentColor;
    }
    .recharts-cartesian-axis-tick-value {
        font-size: 0.875rem;
        fill: currentColor;
    }
    
	.active-dot {
		display: none;
	}
`;
