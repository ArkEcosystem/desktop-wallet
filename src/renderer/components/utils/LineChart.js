import Chart from 'chart.js'
import { Line, mixins } from 'vue-chartjs'
import tailwindConfig from '@tailwind'

Chart.defaults.global.defaultFontFamily = tailwindConfig.fonts.sans.join(',')
// TODO: Add theme colors

export default {
  extends: Line,

  mixins: [mixins.reactiveProp],

  props: {
    chartData: {
      type: Object,
      required: true
    },
    options: {
      type: Object,
      required: true
    }
  },

  mounted () {
    this.$on('chart:render', () => {
      this.$emit('ready')
    })
    this.render()
  },

  watch: {
    options (newOptions) {
      this.destroy()
      this.render()
    }
  },

  methods: {
    getCanvas () {
      return this.$refs.canvas
    },

    destroy () {
      this.$data._chart.destroy()
    },

    render () {
      this.renderChart(this.chartData, this.options)
    },

    update () {
      this.$data._chart.update()
    }
  }
}
