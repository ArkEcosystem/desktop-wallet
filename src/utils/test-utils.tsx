import { render } from "@testing-library/react";
import { i18n } from "app/i18n";
import React from "react";
import { I18nextProvider } from "react-i18next";

const WithProviders = ({ children }: { children: React.ReactNode }) => {
	return <I18nextProvider i18n={i18n}>{children}</I18nextProvider>;
};

const customRender = (component: React.ReactElement, options: any) =>
	render(component, { wrapper: WithProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
