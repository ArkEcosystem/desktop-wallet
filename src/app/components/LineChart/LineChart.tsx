import { SvgCollection } from "app/assets/svg";
import React, { useState } from "react";
import { CartesianGrid, Line, LineChart as RechartsLine, ResponsiveContainer, Tooltip, XAxis, YAxis } from "recharts";
import { styled } from "twin.macro";

import { Icon } from "../Icon";
import { chartStyles } from "./LineChart.styles";

type LineChartProps = {
	data: any[];
	lines: any[];
	period?: string;
	onPeriodClick?: () => void;
	width?: number;
	height?: number;
};

const ActiveDotSvg = SvgCollection["ChartActiveDot"];
const ActiveDot = ({ cx, cy, color }: any) => {
	return (
		<ActiveDotSvg
			className={`text-theme-${color}`}
			width={50}
			height={50}
			x={cx - 25}
			y={cy - 25}
			fill="currentColor"
			stroke="currentColor"
		/>
	);
};

const Dot = ({ cx, cy, index }: any) => {
	return (
		<circle
			cy={cy}
			cx={cx}
			r="5"
			data-testid={`line-chart-dot-${index}`}
			type="monotone"
			stroke="currentColor"
			fill="#FFFFFF"
			className="recharts-dot recharts-l2ne-dot"
			strokeWidth="3"
		/>
	);
};

const ChartLegend = ({ legend = {}, lines, period, onPeriodClick }: any) => {
	return (
		<div>
			<div className="flex">
				{period && (
					<div
						className="pt-4 text-sm font-semibold cursor-pointer text-theme-neutral-700"
						onClick={onPeriodClick}
					>
						<div className="flex">
							<div className="my-auto">Period: {period}</div>
							<div className="my-auto">
								<Icon name="ChevronDown" />
							</div>
						</div>
					</div>
				)}
				<div className="flex justify-end flex-1">
					{legend && <div className="my-auto ml-3 text-sm text-theme-neutral-600">{legend?.label}</div>}
					{lines &&
						lines.map((item: any, index: number) => {
							return (
								<div key={index} className="w-32 p-4 pt-4 pr-0 ml-3 text-right">
									<div
										className={`mr-2 mb-1 border-2 rounded-full w-2 h-2 inline-block align-middle border-theme-${item.color}`}
									/>
									<div className="inline-block text-sm font-semibold text-theme-neutral-700">
										{legend.formatted[item.dataKey] && (
											<span>{legend.formatted[item.dataKey]} - </span>
										)}
										{item.label}
									</div>
								</div>
							);
						})}
				</div>
			</div>
		</div>
	);
};

const Wrapper = styled.div`
	${chartStyles}
`;

export const ChartContent = ({ period, onPeriodClick, data, lines, width, height }: LineChartProps) => {
	const defaultValue = data.concat().pop();
	const [legend, setLegend] = useState(defaultValue);

	return (
		<Wrapper data-testid="line-chart-wrapper" className="text-theme-neutral-300 pb">
			<ChartLegend legend={legend} lines={lines} period={period} onPeriodClick={onPeriodClick} />
			<RechartsLine
				width={width}
				height={height}
				onMouseMove={({ activePayload = [] }) => {
					const { payload } = activePayload[0] || {};
					setLegend(Object.assign({}, defaultValue, payload));
				}}
				data={data}
				margin={{ top: 0, bottom: 0, left: 26, right: 0 }}
			>
				<XAxis dataKey="name" axisLine={false} tick={{ fill: "currentcolor" }} tickSize={20} minTickGap={40} />
				<YAxis axisLine={false} tick={{ fill: "currentColor" }} tickSize={40} />
				<CartesianGrid stroke="currentColor" />
				<Tooltip content={() => <div />} />
				{lines &&
					lines.map((line: any, index: number) => {
						return (
							<Line
								type="monotone"
								key={index}
								dataKey={line.dataKey}
								stroke="currentColor"
								fill="currentColor"
								className={`text-theme-${line.color}`}
								strokeWidth={3}
								yAxisId={0}
								activeDot={<ActiveDot {...line} />}
								dot={<Dot {...line} {...index} />}
							/>
						);
					})}
			</RechartsLine>

			<div className="active-dot">
				<ActiveDot cy={0} cx={0} />
			</div>
		</Wrapper>
	);
};

export const LineChart = (props: LineChartProps) => {
	if (props.width) return <ChartContent {...props} />;
	return (
		<ResponsiveContainer height={props.height} className="mb-16">
			<ChartContent {...props} />
		</ResponsiveContainer>
	);
};

LineChart.defaultProps = {
	data: [],
	lines: [],
	height: 300,
};
