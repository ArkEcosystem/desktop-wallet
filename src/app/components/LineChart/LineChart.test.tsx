import React from "react";
import { act } from "react-dom/test-utils";
import { fireEvent, render, waitFor } from "test-utils";

import { LineChart } from "./LineChart";

describe("LineChart", () => {
	const data = [
		{ name: "Mmn", label: "22 Jun, 2020", usd: 1000, btc: 2400, formatted: { usd: "1,000", btc: "0.26" } },
	];

	const lines = [
		{
			dataKey: "btc",
			label: "BTC",
			color: "warning-600",
		},
	];

	const period = "22 Jun - 28 Jun";

	it("should render", () => {
		const { container } = render(<LineChart data={[]} lines={[]} />);
		expect(container).toMatchSnapshot();
	});

	it("should render with data", () => {
		const { container } = render(<LineChart period={period} data={data} lines={lines} />);
		expect(container).toMatchSnapshot();
	});

	it("should render chart dots", async () => {
		const { getByTestId } = render(
			<div>
				<LineChart width={200} period={period} data={data} lines={lines} />
			</div>,
		);
		const chartWrapper = getByTestId("line-chart-wrapper");
		expect(chartWrapper).toBeTruthy();
		await waitFor(() => expect(getByTestId("line-chart-dot-0")).toBeTruthy(), { timeout: 1000 });
		act(() => {
			fireEvent.mouseMove(chartWrapper.children[1].children[0]);
		});
	});
});
