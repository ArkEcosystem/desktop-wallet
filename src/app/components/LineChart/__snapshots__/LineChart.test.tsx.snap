// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChart should render 1`] = `
<div>
  <div
    class="recharts-responsive-container mb-16"
    id="undefined"
    style="width: 100%; height: 300px;"
  >
    <div
      class="sc-AxirZ fwlMBT text-theme-neutral-300 pb"
      data-testid="line-chart-wrapper"
    >
      <div>
        <div
          class="flex"
        >
          <div
            class="flex justify-end flex-1"
          >
            <div
              class="my-auto ml-3 text-sm text-theme-neutral-600"
            />
          </div>
        </div>
      </div>
      <div
        class="active-dot"
      >
        <svg
          class="text-theme-undefined"
          fill="currentColor"
          height="50"
          stroke="currentColor"
          width="50"
          x="-25"
          y="-25"
        >
          chart-active-dot.svg
        </svg>
      </div>
    </div>
    <div />
  </div>
</div>
`;

exports[`LineChart should render with data 1`] = `
<div>
  <div
    class="recharts-responsive-container mb-16"
    id="undefined"
    style="width: 100%; height: 300px;"
  >
    <div
      class="sc-AxirZ fwlMBT text-theme-neutral-300 pb"
      data-testid="line-chart-wrapper"
    >
      <div>
        <div
          class="flex"
        >
          <div
            class="pt-4 text-sm font-semibold cursor-pointer text-theme-neutral-700"
          >
            <div
              class="flex"
            >
              <div
                class="my-auto"
              >
                Period: 
                22 Jun - 28 Jun
              </div>
              <div
                class="my-auto ml-1"
              >
                <div
                  class="sc-AxjAm cpczEI"
                  height="1em"
                  width="1em"
                >
                  <svg>
                    chevron-down.svg
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <div
            class="flex justify-end flex-1"
          >
            <div
              class="my-auto ml-3 text-sm text-theme-neutral-600"
            >
              22 Jun, 2020
            </div>
            <div
              class="w-32 p-4 pt-4 pr-0 ml-3 text-right"
            >
              <div
                class="mr-2 mb-1 border-2 rounded-full w-2 h-2 inline-block align-middle border-theme-warning-600"
              />
              <div
                class="inline-block text-sm font-semibold text-theme-neutral-700"
              >
                <span>
                  0.26
                   - 
                </span>
                BTC
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="active-dot"
      >
        <svg
          class="text-theme-undefined"
          fill="currentColor"
          height="50"
          stroke="currentColor"
          width="50"
          x="-25"
          y="-25"
        >
          chart-active-dot.svg
        </svg>
      </div>
    </div>
    <div />
  </div>
</div>
`;
